packages/punchcard/lib/appsync/vtl-object.ts

```diff
@@ -0,0 +1,208 @@
+  export type TypeOf<T extends VObject> = T[typeof type];
+
+  export function isObject(a: any): a is VObject {
+    return a[expr] !== undefined;
+  }
+
+  // export type ShapeOf<T extends VObject> = T extends VObject<infer I> ? I : never;
+
+  export type Of<T extends Shape> =
+    T extends RecordShape<infer M> ? VRecord<{
+      [m in keyof M]: Of<M[m]>;
+    }> & {
+      [m in keyof M]: Of<M[m]>;
+    } :
+    T extends ArrayShape<infer I> ? VList<VObject.Of<I>> :
+    T extends MapShape<any> ? never : // maps are not supported in GraphQL
+    T extends BoolShape ? VBool :
+    T extends DynamicShape<any> ? VAny :
+    T extends IntegerShape ? VInteger :
+    T extends NumberShape ? VFloat :
+    T extends StringShape ? VString :
+    T extends TimestampShape ? VTimestamp :
+
+    VObject<T>
+    ;
+
+  /**
+   * Object that is "like" a VObject for some Shape.
+   *
+   * Like meaning that is either an expression, or a collection
+   * of expressions that share the structure of the target type.
+   */
+  export type Like<T extends Shape> = Value.Of<T> | VObject.Of<T> | (
+    T extends RecordShape<infer M> ? {
+      [m in keyof M]: Like<Pointer.Resolve<M[m]>>;
+    } :
+    T extends ArrayShape<infer I> ? Like<I>[] :
+    T extends SetShape<infer I> ? Like<I>[] :
+    T extends MapShape<infer I> ? {
+      [key: string]: Like<I>;
+    } :
+    VObject.Of<T> | Value.Of<T>
+  );
+}
+
+export const IDTrait: {
+  [Trait.Data]: {
+    graphqlType: 'ID'
+  }
+} = {
+  [Trait.Data]: {
+    graphqlType: 'ID'
+  }
+};
+
+export const ID = string.apply(IDTrait);
+
+export class VAny extends VObject<AnyShape> {}
+export class VUnknown extends VObject<UnknownShape> {}
+export class VInteger extends VObject<IntegerShape> {}
+export class VFloat extends VObject<NumberShape> {}
+export class VNothing extends VObject<NothingShape> {}
+export class VNever extends VObject<NeverShape> {}
+export class VBinary extends VObject<BinaryShape> {}
+
+export class VBool extends VObject<BoolShape> {
+  public static not(a: VBool): VBool {
+    return new VBool(bool, VObject.exprOf(a).prepend('!'));
+  }
+}
+
+export class VString extends VObject<StringShape> {
+  public toUpperCase(): VString {
+    return new VString(VObject.typeOf(this), VObject.exprOf(this).dot('toUpperCase()'));
+  }
+
+  public isNotEmpty(): VBool {
+    return VBool.not(this.isEmpty());
+  }
+
+  public isEmpty(): VBool {
+    return new VBool(bool, VObject.exprOf(this).dot('isEmpty()'));
+  }
+
+  public size(): VInteger {
+    return new VInteger(integer, VObject.exprOf(this).dot('size()'));
+  }
+}
+
+export class VTimestamp extends VObject<TimestampShape> {}
+
+export class VList<T extends VObject = VObject> extends VObject<ArrayShape<VObject.TypeOf<T>>> {
+  constructor(shape: ArrayShape<VObject.TypeOf<T>>, expression: VExpression) {
+    super(shape, expression);
+  }
+}
+
+
+export class VSet<T extends VObject = VObject> extends VObject<SetShape<VObject.TypeOf<T>>> {
```

Comment:
At first glance, this has me thing about how there is no `Set` type in GraphQL. What'm I missing here?